### PR TITLE
WIP: Fix circular

### LINKED ISSUE #69: Stack overflow on parsing (large) OpenAPI 2.0 specification
Hi, 
We've had a user of our OpenAPI documentation tool [DapperDox](https://github.com/DapperDox/dapperdox) report a Stack Overflow with our latest release. This was built today using master of go-openapi.

From the log, the stack overflow appears to be caused somewhere in `go-openapi/spec/expander.go` and we _assume_ it may be caused by the size of the specification, as it is quite big, at 84K lines.

He has uploaded his specification and error log giving the stacktrace in a DapperDox issue:
[DapperDox Issue - Stack Overflow](https://github.com/DapperDox/dapperdox/issues/83)

To re-cook the binary releases, we have temporarily wound `go-openapi/spec` back to commit [e51c28f07](https://github.com/go-openapi/spec/commit/e51c28f07047ad90caff03f6450908720d337e0c) and `go-openapi/analysis` back to [7427b3df9](https://github.com/go-openapi/analysis/commit/7427b3df9eab2fa6ffe94659dd42268847562803), which was stable for us previously.

Are you aware of issues with large specifications?

Cheers!

### LINKED ISSUE #75: spec.ExpandSpec doesn't seem to terminate in some swagger spec with recursive references
Hi,
I noticed this problem when generating a go client from one of the odata-openapi samples. The generation gets stuck and looking into this problem revealed that the call to spec.EpandSpec is not coming back. While looking at the recursive iterations that go into expandSchema in expander.go, I am wondering why we are not doing something like isSeen check in addition to isCircular check.

I put a test case that shows this issue at my forked repo.
https://github.com/elakito/spec/blob/issue-expand/expander_test.go#L454
and this is the swagger file used.
https://github.com/elakito/spec/blob/issue-expand/fixtures/expansion/circularSpec2.json

If someone can comment on this issue, that would be appreciated.

regards, aki

### LINKED ISSUE #76: spec ExpandSpec on circular $ref leaves absolute path in new $ref
When running expand on a spec with a circular $ref, the resulting $ref is left with the full absolute path of the file.

Example: https://github.com/go-swagger/go-swagger/fixtures/canary/bitbucket.org/swagger.json

There are two circular $ref in this spec: 
```
#/definitions/repository
#/definitions/base_commit
```

If I use expand fromthe `swagger expand` command, or independently from `go-openapi/analysis/flatten(), the resolved circular $ref leaves the full absolute path of the file, instead of just `#/definitions/...`

In both case, the `RelativeBase` expand option parameter is set to the absolute path of the spec file.

This is annoying when in go-openapi/analysis, we try to sort out $ref's which are plain definitions.

### PR DESCRIPTION
**Pull request by pytlesk4**
Fixes #79 
Fixes #76 
Fixes #75 (with some limitations, see #75 and related new issues)
Fixes #69

### CODE REVIEW

File: expander.go [hunk 1]
```diff
@@ -627,8 +651,20 @@ func basePathFromSchemaID(oldBasePath, id string) string {
 	return u.String()
 }
 
-func isCircular(ref *Ref, basePath string, parentRefs ...string) bool {
-	return basePath != "" && swag.ContainsStringsCI(parentRefs, ref.String())
+func (r *schemaLoader) isCircular(ref *Ref, basePath string, parentRefs ...string) (foundCycle bool) {
+	normalizedRef := normalizePaths(ref.String(), basePath)
+	r.context.Lock()
```
[the remark this answers is not in the document]
Comment (pytlesk4) [reply expander.go]: In the validate package, test were failing for accessing the context concurrently.

File: expander.go [hunk 2]
```diff
@@ -683,12 +720,20 @@ func expandSchema(target Schema, parentRefs []string, resolver *schemaLoader, ba
...
-			resolver, err = transitiveResolver(basePath, target.Ref, resolver)
+			transitiveResolver, err := transitiveResolver(basePath, target.Ref, resolver)
 			if shouldStopOnError(err, resolver.options) {
 				return nil, err
 			}
 
-			return expandSchema(*t, parentRefs, resolver, normalizedBasePath)
+			if transitiveResolver != resolver {
+				if transitiveResolver.options != nil && transitiveResolver.options.RelativeBase != "" {
+					basePath, _ = absPath(transitiveResolver.options.RelativeBase)
```
[the remark this answers is not in the document]
Comment (pytlesk4) [reply expander.go]: Correct, good catch, the relative base path should def be changed.

[the remark this answers is not in the document]
Comment (fredbi) [reply expander.go]: I am finishing up the integration tests with analysis with my new commit, validate and the full go-swagger CI.
Looks like it is running well so far (it has not completed yet)...

File: expander.go (hunk 1 above)
[the remark this answers is not in the document]
Comment (fredbi) [reply expander.go]: how come is that? I just run the full suite without problem after I removed this lock.

I don't see the logics in locking: the context is attached to the schemaLoader object...
Please advise

File: expander.go (hunk 2 above)
[the remark this answers is not in the document]
Comment (fredbi) [reply expander.go]: ok go-swagger full CI scripts did complete ok on my machine.

File: expander.go [hunk 3]
```diff
@@ -67,14 +62,33 @@ func initResolutionCache() ResolutionCache {
 	}}
 }
 
+// resolverContext allows to share a context during spec processing.
+// At the moment, it just holds the index of circular references found.
+type resolverContext struct {
+	// circulars holds all visited circular references, which allows shortcuts.
+	// NOTE: this is not just a performance improvement: it is required to figure out
```
Comment (fredbi) [inline expander.go]: I removed your lock here. Please advise. I don't understand how come we could have a concurrency issue here.
On my side I ran the validate tests without problem.

Comment (pytlesk4) [reply expander.go]: I just tried to re run the tests too, and couldn't reproduce it. I swear it was happening last night, so strange.

File: expander.go [hunk 4]
```diff
@@ -369,6 +382,59 @@ func normalizePaths(refPath, base string) string {
 	return baseURL.String()
 }
 
+// denormalizePaths returns to simplest notation on file $ref,
+// i.e. strips the absolute path and sets a path relative to the base path.
+//
+// This is currently used when we rewrite ref after a circular ref has been detected
+func denormalizeFileRef(ref *Ref, relativeBase, originalRelativeBase string) *Ref {
```
Comment (fredbi) [inline expander.go]: This is more complicated than expected at first. I tested it against JSONSchema schema, swagger schema to make sure resolved $ref are okay.

**Comment by pytlesk4** [comment]
This looks great.

File: expander.go [hunk 5]
```diff
@@ -904,7 +995,13 @@ func transitiveResolver(basePath string, ref Ref, resolver *schemaLoader) (*sche
 		rootURL.Fragment = ""
 		root, _ := resolver.cache.Get(rootURL.String())
 		var err error
-		resolver, err = defaultSchemaLoader(root, resolver.options, resolver.cache)
+
+		// shallow copy of resolver options to set a new RelativeBase when
```
Comment (fredbi) [inline expander.go]: This is the new transitive: changing root with a shallow clone of options to set a new relative base

**Comment by fredbi** [comment]
ok
I let you pass your tests. No rush to merge.
On my side, I am pushing a vendor update PR on go-swagger.

**Comment by pytlesk4** [comment]
Looks good to merge on my end.

**Comment by fredbi** [comment]
Just found another little glitch when opts is passed a relative path... So I need one more commit.
Also acknowledge #69 fixed (was also circular related)

**Pull request merged**